File: optimize_hero.py
#!/usr/bin/env python3
"""
Script to optimize the hero image:
- Convert to WebP format
- Compress to target size < 300KB
- Maintain good visual quality
"""

from PIL import Image
import os

def optimize_hero_image(input_path, output_webp, target_size_kb=280, quality_start=85):
    """
    Optimize image to WebP format with target file size.

    Args:
        input_path: Path to input JPEG
        output_webp: Path for output WebP
        target_size_kb: Target file size in KB
        quality_start: Starting quality level (0-100)
    """
    print(f"Opening image: {input_path}")
    img = Image.open(input_path)

    # Get original dimensions
    original_size = os.path.getsize(input_path) / 1024
    print(f"Original size: {original_size:.2f} KB ({img.size[0]}x{img.size[1]})")

    # Start with high quality and reduce until we hit target size
    quality = quality_start

    while quality > 50:
        # Save with current quality
        img.save(output_webp, 'WEBP', quality=quality, method=6)

        # Check file size
        webp_size = os.path.getsize(output_webp) / 1024

        print(f"Quality {quality}: {webp_size:.2f} KB")

        if webp_size <= target_size_kb:
            print(f"✓ Target achieved! Final size: {webp_size:.2f} KB at quality {quality}")
            reduction = ((original_size - webp_size) / original_size) * 100
            print(f"✓ Size reduction: {reduction:.1f}%")
            return True

        quality -= 5

    print(f"⚠ Could not reach target size. Final: {webp_size:.2f} KB")
    return False

if __name__ == "__main__":
    input_file = "photos/pro05.jpg"
    output_file = "photos/pro05.webp"

    if not os.path.exists(input_file):
        print(f"Error: {input_file} not found!")
        exit(1)

    optimize_hero_image(input_file, output_file, target_size_kb=280, quality_start=85)
    print(f"\n✓ WebP image created: {output_file}")

File: optimize_hero.js
#!/usr/bin/env node
/**
 * Script to optimize the hero image:
 * - Convert to WebP format
 * - Compress to target size < 300KB
 * - Maintain good visual quality
 */

const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

async function optimizeHeroImage(inputPath, outputWebp, targetSizeKB = 280) {
    console.log(`Opening image: ${inputPath}`);

    // Get original file size
    const originalStats = fs.statSync(inputPath);
    const originalSizeKB = originalStats.size / 1024;
    console.log(`Original size: ${originalSizeKB.toFixed(2)} KB`);

    // Get image metadata
    const metadata = await sharp(inputPath).metadata();
    console.log(`Dimensions: ${metadata.width}x${metadata.height}`);

    // Try different quality levels
    let quality = 85;
    let bestQuality = 85;
    let bestSize = Infinity;

    while (quality >= 50) {
        await sharp(inputPath)
            .webp({ quality: quality, effort: 6 })
            .toFile(outputWebp);

        const stats = fs.statSync(outputWebp);
        const webpSizeKB = stats.size / 1024;

        console.log(`Quality ${quality}: ${webpSizeKB.toFixed(2)} KB`);

        if (webpSizeKB <= targetSizeKB) {
            console.log(`✓ Target achieved! Final size: ${webpSizeKB.toFixed(2)} KB at quality ${quality}`);
            const reduction = ((originalSizeKB - webpSizeKB) / originalSizeKB) * 100;
            console.log(`✓ Size reduction: ${reduction.toFixed(1)}%`);
            return true;
        }

        bestSize = webpSizeKB;
        bestQuality = quality;
        quality -= 5;
    }

    console.log(`⚠ Could not reach target size. Final: ${bestSize.toFixed(2)} KB at quality ${bestQuality}`);
    return false;
}

// Run optimization
const inputFile = path.join(__dirname, 'photos/pro05.jpg');
const outputFile = path.join(__dirname, 'photos/pro05.webp');

if (!fs.existsSync(inputFile)) {
    console.error(`Error: ${inputFile} not found!`);
    process.exit(1);
}

optimizeHeroImage(inputFile, outputFile, 280)
    .then(() => {
        console.log(`\n✓ WebP image created: ${outputFile}`);

        // Show final comparison
        const originalSize = fs.statSync(inputFile).size / 1024;
        const webpSize = fs.statSync(outputFile).size / 1024;
        console.log(`\nComparison:`);
        console.log(`  Original JPEG: ${originalSize.toFixed(2)} KB`);
        console.log(`  Optimized WebP: ${webpSize.toFixed(2)} KB`);
        console.log(`  Savings: ${((1 - webpSize/originalSize) * 100).toFixed(1)}%`);
    })
    .catch(err => {
        console.error('Error:', err);
        process.exit(1);
    });
